### ISSUE TITLE
Backup file again when data blobs are missing

### ISSUE THREAD

**Issue by MichaelEischer** [restic/restic#2571]
Output of `restic version`
--------------------------
0.9.6-dev built from 0f22f00

How did you run restic exactly?
-------------------------------
```
restic -r test-repo backup restic
# damage the repository on purpose
rm test-repo/<datafile in repo>
# recover by rebuilding the index
restic -r test-repo rebuild-index
restic -r test-repo backup restic
restic -r test-repo check
```
```
using temporary cache in /var/folders/vs/zdx8nt1n4xlc2fgvtw589_100000gn/T/restic-check-cache-322660528
repository 115e9410 opened successfully, password is correct
created new cache in /var/folders/vs/zdx8nt1n4xlc2fgvtw589_100000gn/T/restic-check-cache-322660528
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
error for tree 0b6d095c:
  tree 0b6d095c: file "restic" blob 6 size could not be found
  tree 0b6d095c: file "restic" blob 10 size could not be found
  tree 0b6d095c: file "restic" blob 11 size could not be found
  tree 0b6d095c: file "restic" blob 12 size could not be found
  tree 0b6d095c: file "restic" blob 15 size could not be found
  tree 0b6d095c: file "restic" blob 17 size could not be found
  tree 0b6d095c, blob 7d932deb: not found in index
  tree 0b6d095c, blob 238d45b0: not found in index
  tree 0b6d095c, blob 22038c9a: not found in index
  tree 0b6d095c, blob 2c3e7399: not found in index
  tree 0b6d095c, blob db1787f2: not found in index
  tree 0b6d095c, blob 1b0e9af3: not found in index
```

Expected behavior
-----------------
Restic should automatically recover the missing data blob when coming across them in a later backup run.

Actual behavior
---------------
Data blobs are still missing.

Do you have any idea what may have caused this?
-----------------------------------------------
The new archiver code just copies the blob list:
https://github.com/restic/restic/blob/0f22f008ede3e46b184cdedcd8c16e69d1969b57/internal/archiver/archiver.go#L388

The old archiver implementation did check the whole blob list beforehand:
https://github.com/restic/restic/blob/3cd92efdcfccd09bf69fadccb2f6f4fa500577e1/internal/archiver/archiver.go#L304-L311

I couldn't find an issue or commit message that states this change was made intentionally or not.

Do you have an idea how to solve the issue?
-------------------------------------------
I've implemented the necessary changes in https://github.com/MichaelEischer/restic/tree/recover-blobs-on-backup . So far I did not benchmark how expensive the additional index lookups are.
The alternative would be to tell everyone who encountered missing blobs to run `rebuild-index` and `backup --force` afterwards.

**Comment by fd0** [comment]
I don't think this change was intentional when I rewrote the archiver. I also think it'd be a great addition :)

**Comment by MichaelEischer** [comment]
Using `go version go1.14.2 darwin/amd64` this added 6 seconds cpu time (according to the cpu profile) to a backup run with 450k files and 320GiB size, which due to mostly unchanged files took 2:18 in total.

And for a large 8.5TiB repo with 62 million blobs backuping a folder with 330GB and 1 million files increased the time from 16:47 to 18:24 for a backup run with no changed files (using `GOMAXPROCS=1`). Both runs take (CPU time according to the cpu profile) roughly 9 minutes to load the index and 1.5 minutes to scan the filesystem. The backup run itself (Archiver.Snapshot) slows down from 193s to 306s.

So in total the performance hit becomes noticeable for large repositories (multi-TB) as the index lookups is rather slow for those, but is not too horrible.

**Comment by Mikescher** [comment]
@MichaelEischer: Hey thanks man - this really saved my ass just now.

Due to a hard drive failure I damaged a single file in the /data/ directory.  
Fortunately I still had the original file and with the version from your branch a new backup would recreate the missing data file.

For reference here is what I did:
~~~
> rm data/{broken_file}
> rm index/{broken_file}      # an index file was broken too
> restic check
> restic rebuild-index
> restic check                # this on told me which snapshots (all) and which files (only two) were affected
> restic backup               # created a new backup re-adding the two (now-missing) files
> restic check                # all looks good now
~~~

I don't know if you will get this merged into master but at least for me it was really helpful - perhaps it can be disabled by a parameter, or it only runs when the last `check` failed?

&nbsp;

Regards *(any many thanks)*  
  ~ Mike

**Comment by MichaelEischer** [comment]
Happy to hear that my draft helped :-) . `backup --force` should also be able to recover missing blobs, but of course it has to reread all files in the backup source.

I hope to complete preparing the PR this weekend (its mostly missing tests right now).

As restic doesn't remember problems between command runs, it won't be possible to just enable this when `check` failed. If always checking that all blobs still exist turns out to be too slow, we could still add a option to disable it. However, the cleanest solution would be to just fix the performance of the main index, which is currently way to slow for large repositories.

**Comment by aawsome** [comment]
> So in total the performance hit becomes noticeable for large repositories (multi-TB) as the index lookups is rather slow for those, but is not too horrible.

This should be cured by a merged index, like #2818 .

**Comment by jerith666** [comment]
Just a note that I applied this patch locally as well, and it recovered the `not found` errors I was seeing in my repo.  FWIW, I had previously tried `backup --force` and that did _not_ repair the problem.
